### ISSUE TITLE
Prevent DynamicMap from running callback on initialization?

### ISSUE THREAD

**Issue by nikhilxb** [ioam/holoviews#1729]
TLDR: What is the best way to use dynamic maps when the requisite data is not yet available? This seems like a common need for dynamic plotting (e.g. I need dynamic plotting for the displaying in real-time the training/validation loss for machine learning models). 
Right now, `dmap = hv.DynamicMap(update_fn, ...)` calls `update_fn`, which is problematic if there is no data to plot. What should `update_fn` return if it can't plot anything?

---
Here's an example that mirrors the problem I'm facing:
I want to generate a dynamic plot of two functions, plotting 10 samples each at a time:
- squared `x**2` for `x` in [1, 10)
- cubed `x**3` for `x` in [1, 10)
- squared `x**2` for `x` in [10, 20)
- cubed `x**3` for `x` in [10, 20)

The data is generated in an object, which must be retrieved inside `update_fn`. Since this example snippet assumes a notebook, I'll mimic this with a global var in another cell.

`In [0]:`
```python
import time
import holoviews as hv
import holoviews.streams as hvstrm
hv.extension("bokeh")
```
`In [1]:`
```python
%%opts Curve {+framewise} [width=1200 height=600] (line_width=3)

def update_fn():

    # PROBLEM 1
    global points  # points = {"xsq":[...], "ysq":[...], "xcu":[...], "ycu":[...]}
    
    # PROBLEM 2
    if (not points["xsq"]) and (not points["xcu"]):
        disp = hv.Curve(([0], [0]), kdims=["x0"], vdims=["y0"])
        return disp
    
    # Display squared (sq)
    if(points["xsq"]):
        dispsq = hv.Curve((points["xsq"], points["ysq"]), kdims=["x"], vdims=["squared"])
        dispsq.redim.range(x=(0, points["xsq"][-1]), squared=(0, points["ysq"][-1]))
        disp = dispsq
    
    # Display cubed (cu)
    if(points["xcu"]):
        dispcu = hv.Curve((points["xcu"], points["ycu"]), kdims=["x"], vdims=["cubed"])
        dispcu.redim.range(x=(0, points["xcu"][-1]), squared=(0, points["ycu"][-1]))
        disp *= dispcu
    
    return disp

dmap = hv.DynamicMap(update_fn, streams=[hvstrm.Stream.define("Dummy")()]);
dmap
```
`In [2]:`
```python
points = {"xsq":[], "ysq":[], "xcu":[], "ycu":[]}  # Declaration line

for n in range (0, 2):
    
    # Squared
    for x in range(0, 10):
        x += 10*n
        points["xsq"].append(x)
        points["ysq"].append(x**2)
        dmap.event()
        time.sleep(0.5)

    # Cubed
    for x in range(0, 10):
        x += 10*n
        points["xcu"].append(x)
        points["ycu"].append(x**3)
        dmap.event()
        time.sleep(0.5)
```
PROBLEM 1 occurs because `update_fn` is called upon creation of the `DynamicMap`, before the points have been initialized:
`Out [1]:`
```
NameError: name 'points' is not defined
```
But, even if `points` is defined (declaration line moved to cell 1), I run into PROBLEM 2. This would seem like an workable (albeit hacky) way of displaying an empty plot before data is available. But this still runs into issues:
`Out [2]:`
```
/usr/local/miniconda3/lib/python3.6/site-packages/bokeh/models/sources.py:91: BokehUserWarning: ColumnDataSource's columns must be of the same length.
Current lengths: ('squared', 2), ('x', 2), ('x0', 1), ('y0', 1)
  "Current lengths: %s" % ", ".join(sorted(str((k, len(v))) for k, v in data.items())), BokehUserWarning))
/usr/local/miniconda3/lib/python3.6/site-packages/bokeh/models/sources.py:91: BokehUserWarning: ColumnDataSource's columns must be of the same length.
Current lengths: ('squared', 3), ('x', 3), ('x0', 1), ('y0', 1)
  "Current lengths: %s" % ", ".join(sorted(str((k, len(v))) for k, v in data.items())), BokehUserWarning))
/usr/local/miniconda3/lib/python3.6/site-packages/bokeh/models/sources.py:91: BokehUserWarning: ColumnDataSource's columns must be of the same length.
Current lengths: ('squared', 4), ('x', 4), ('x0', 1), ('y0', 1)
  "Current lengths: %s" % ", ".join(sorted(str((k, len(v))) for k, v in data.items())), BokehUserWarning))
...
/usr/local/miniconda3/lib/python3.6/site-packages/bokeh/models/sources.py:91: BokehUserWarning: ColumnDataSource's columns must be of the same length.
Current lengths: ('squared', 10), ('x', 10), ('x0', 1), ('y0', 1)
  "Current lengths: %s" % ", ".join(sorted(str((k, len(v))) for k, v in data.items())), BokehUserWarning))
/usr/local/miniconda3/lib/python3.6/site-packages/bokeh/models/sources.py:91: BokehUserWarning: ColumnDataSource's columns must be of the same length.
Current lengths: ('squared', 10), ('x', 11), ('x0', 1), ('y0', 1)
  "Current lengths: %s" % ", ".join(sorted(str((k, len(v))) for k, v in data.items())), BokehUserWarning))
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
<ipython-input-3-cc0b44e54a4e> in <module>()
     14         points["xcu"].append(x)
     15         points["ycu"].append(x**3)
---> 16         dmap.event()
     17         time.sleep(0.5)

/usr/local/miniconda3/lib/python3.6/site-packages/holoviews/core/spaces.py in event(self, **kwargs)
    811             stream.update(**rkwargs)
    812 
--> 813         Stream.trigger(self.streams)
    814 
    815     def _style(self, retval):

/usr/local/miniconda3/lib/python3.6/site-packages/holoviews/streams.py in trigger(cls, streams)
    146         with triggering_streams(streams):
    147             for subscriber in subscribers:
--> 148                 subscriber(**dict(union))
    149 
    150         for stream in streams:

/usr/local/miniconda3/lib/python3.6/site-packages/holoviews/plotting/plot.py in refresh(self, **kwargs)
    505         # Update if not top-level, batched or an ElementPlot
    506         if not self.top_level or isinstance(self, GenericElementPlot):
--> 507             self.update(stream_key)
    508 
    509         if self.comm is not None and self.top_level:

/usr/local/miniconda3/lib/python3.6/site-packages/holoviews/plotting/plot.py in update(self, key)
    484         if len(self) == 1 and key == 0 and not self.drawn:
    485             return self.initialize_plot()
--> 486         item = self.__getitem__(key)
    487         self.traverse(lambda x: setattr(x, '_updated', True))
    488         return item

/usr/local/miniconda3/lib/python3.6/site-packages/holoviews/plotting/plot.py in __getitem__(self, frame)
    222         if not isinstance(frame, tuple):
    223             frame = self.keys[frame]
--> 224         self.update_frame(frame)
    225         return self.state
    226 

/usr/local/miniconda3/lib/python3.6/site-packages/holoviews/plotting/bokeh/element.py in update_frame(self, key, ranges, plot, element, empty)
    863         plot = self.handles['plot']
    864         if not self.overlaid:
--> 865             self._update_ranges(style_element, ranges)
    866             self._update_plot(key, plot, style_element)
    867 

/usr/local/miniconda3/lib/python3.6/site-packages/holoviews/plotting/bokeh/element.py in _update_ranges(self, element, ranges)
    549         l, b, r, t = None, None, None, None
    550         if any(isinstance(r, (Range1d, DataRange1d)) for r in [x_range, y_range]):
--> 551             l, b, r, t = self.get_extents(element, ranges)
    552             if self.invert_axes:
    553                 l, b, r, t = b, l, t, r

ValueError: not enough values to unpack (expected 4, got 0)
```

**Comment by philippjfr** [comment]
@nikhilxb There are multiple options here but I may have to clarify what exactly you're trying to achieve a little bit. As far as I can tell this achieves what you want:

```python
%%opts Curve {+framewise} [width=800 height=600] (line_width=3)

points = {"xsq":[], "ysq":[], "xcu":[], "ycu":[]}  # Declaration line

def update_fn():
    global points 
    dispsq = hv.Curve((points["xsq"], points["ysq"]), kdims=["x"], vdims=["squared"])
    dispcu = hv.Curve((points["xcu"], points["ycu"]), kdims=["x"], vdims=["cubed"])
    return dispsq * dispcu

dmap = hv.DynamicMap(update_fn, streams=[hv.streams.Stream.define("Dummy")()]);
dmap

And then in another cell:


```
for n in range (0, 2):
    # Squared
    for x in range(0, 10):
        x += 10*n
        points["xsq"].append(x)
        points["ysq"].append(x**2)
        dmap.event()
        time.sleep(0.5)

    # Cubed
    for x in range(0, 10):
        x += 10*n
        points["xcu"].append(x)
        points["ycu"].append(x**3)
        dmap.event()
        time.sleep(0.5)
```

Maybe I'm misunderstanding what you are trying to achieve however. Secondly here's several tips:

* A DynamicMap callback should generally always return the same number and types of Elements.
* Most Elements allow declaring empty data with an empty list, e.g. ``hv.Curve([], , kdims=["x"], vdims=["cubed"])``

I think the main issue you ran into was the fact you variously returned a single Element or multiple Elements. Let me know if I've misunderstood your problem or the solution I've provided does not address your needs. It's likely that we need to make this requirement clearer in our documentation. That said I have plans to eventually allow dynamically adding or removing new Elements in a callback but that probably won't happen before 2.0.

**Comment by nikhilxb** [comment]
@philippjfr Your solution worked for me. Great tips -- exactly the 2 things I wasn't getting. Yes, I think more documentation about the requirements about the DynamicMap in particular would be helpful. Thanks!

**Comment by NightFantomJ2** [comment]
@philippjfr it sounds like I have a similar issue but I'm not sure whether this applies, let me know if I need to make a new issue instead of diluting this one.

I have a dynamic number of columns in my dataframe depending on what the user wants to see grouped (basically sql-like group-by clauses), and one boolean value per potential group-by colunm:

```
dmap = hv.DynamicMap(make_insight_curves, kdims=['group_name', 'group_resultCode'])
dmap.redim.values(group_name=(True, False)).redim.values(group_resultCode=(True, False))
```

now this function make_insight_curves gives one or more hv.Curve objects (one per column in the dataframe), nested in an hv.Overlay (I've also tried ndoverlay as per your comment, but this didn't fix it)

From your comment I understand that returning a variable amount of elements doesn't work with dynamicmaps. Is there a different kind of element that could do a similar job?

**Comment by philippjfr** [comment]
>From your comment I understand that returning a variable amount of elements doesn't work with dynamicmaps. Is there a different kind of element that could do a similar job?

Returning a variable number is now allowed.
